flink-runtime/src/main/java/org/apache/flink/runtime/client/JobClientActor.java

```diff
@@ -59,11 +87,50 @@ public JobClientActor(
 		this.leaderSessionID = Preconditions.checkNotNull(leaderSessionID, "The leader session ID option must not be null.");
 
 		this.sysoutUpdates = sysoutUpdates;
+		// set this to -1 to indicate the job hasn't been created yet.
+		this.currentJobCreatedAt = -1;
 	}
 	
 	@Override
 	protected void handleMessage(Object message) {
 		
+		// ======= Job status messages on regular intervals ==============
+		if(message instanceof JobManagerMessages.CurrentJobStatus){
+			JobStatus statusReport = ((JobManagerMessages.CurrentJobStatus) message).status();
+			long timeDiff;
+			switch(statusReport){
+				case RUNNING:
+					// Vincent, we happy?
+					this.currentJobCreatedAt = -1;
+					break;
+				case FINISHED:
+					// Yeah! We happy!
+					this.currentJobCreatedAt = -1;
+					break;
+				case CREATED:
+					// we're still at Job CREATED. Let's see if we're over the limit.
+					timeDiff = (System.currentTimeMillis() - this.currentJobCreatedAt);
+					if(timeDiff > JOB_CLIENT_JOB_STATUS_TIMEOUT){
+						failWithTimeout(timeDiff);
+					} // otherwise just wait a bit longer.
+					break;
+				case RESTARTING:
```

Comment:
No it might happen due to a failure in some external system, e.g. HDFS. The problem is that you keep increasing the `timeDiff` even though it should have been restarted in this case. I guess you would have to register for the job status update messages at the `JobManager` to solve the problem.